File: opencl-intel/cmake_utils/c++_2_c#.pl
#!perl

use File::Basename;

# ------- help ----------
sub help
{
  printf "Usage:
  $0 <input_file> <project-name>

  Replace DevStudio 2008 Solution project type from C++ to C#
  for given project. Original solution is saved with 
  <input_file>.orig name
";
}

# ----- main ----------

my $num_of_args = $#ARGV + 1;
my ($in_file_name, $proj_name) = @ARGV;
my $CXX_UUID = "{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}";
my $C_SHARP_UUID = "{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}";
my $tmp_file_name = "$in_file_name.new";
my $save_file_name = "$in_file_name.orig";

if ($num_of_args != 2)
{
  help();
  exit 1;
}

# try to open files
if (!open (IN_FILE_HANDLE, "<$in_file_name"))
{
    print "Cannot open input file: $in_file_name\n";
    exit 1;
}

if (!open (OUT_FILE_HANDLE, ">$tmp_file_name"))
{
    print "Cannot open output file: $tmp_file_name\n";
    close IN_FILE_HANDLE;
    exit 1;
}

while (<IN_FILE_HANDLE>) 
{
    if (/^Project\(.+\"$proj_name\".+/)
    {
        s/$CXX_UUID/$C_SHARP_UUID/;
    }
    print OUT_FILE_HANDLE;
}


close IN_FILE_HANDLE;
close OUT_FILE_HANDLE;

rename( $in_file_name, $save_file_name );
rename( $tmp_file_name, $in_file_name );

exit 0;

File: opencl-intel/cmake_utils/c++_2_c#.py
#!python

import sys,os,re

def parse():
    if len(sys.argv) < 3 :
        print """
Usage: """, sys.argv[0],""" <.sln-file-name> <project-name-to-convert>

Convert project type from C++ to C# in the Dev Studio Solution 2008 file.
"""
        exit(1)

    file = sys.argv[1]
    proj = sys.argv[2:]
    return (file, proj)


def isPatternsExist(line, patterns):
    for pattern in patterns:
	if pattern.search(line):
	    return True
    return False
	
def main():
    (file, proj) = parse()
    tmpfile_name = file + '.tmp'
    oldfile_name = file + '.orig'
    CXX_UUID = '{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}'
    C_SHARP_UUID = '{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}'

    try:
        infile = open(file, 'r')
    except IOError,(errno, strerror):
        print 'Cannot open input file',file,':',strerror
        sys.exit(1)
    
    try:
        tmpfile = open(tmpfile_name, 'w')
    except IOError,(errno, strerror):
        print 'Cannot open output file',tmpfile_name,':',strerror
        sys.exit(1)

    patterns = []
    for p in proj:
	patterns.append(re.compile('^Project.+"' + p + '".+'))
    for line in infile:
        if isPatternsExist(line, patterns): #pattern.search(line):
            line = re.sub(CXX_UUID,C_SHARP_UUID,line)
        tmpfile.write(line)

    infile.close()
    tmpfile.close()

    try:
        if os.path.isfile( oldfile_name ):
            os.remove(oldfile_name)
    except IOError,(errno, strerror):
        print 'Cannot remove file',oldfile_name,':',strerror
        sys.exit(1)

    try:
        os.rename( file, oldfile_name )
    except WindowsError,(errno, strerror):
        print 'Cannot rename file',file, 'to',oldfile_name,':',strerror
        sys.exit(1)

    try:
        os.rename( tmpfile_name, file )
    except WindowsError,(errno, strerror):
        print 'Cannot rename file',tmpfile_name, 'to',file,':',strerror
        print 'Reverting....'
        os.rename( oldfile_name, file )
        print 'Done'
        sys.exit(1)

# main
if __name__ == "__main__":
    main()
    sys.exit(0)

        
        
        
        
    
